### ISSUE TITLE
NFS shares and volumes don't mix

### ISSUE THREAD

**Issue by cameron** [docker/docker#4213]
Narrowed from https://github.com/dotcloud/docker/issues/4142

When running the docker daemon inside a virtual machine that uses NFS to share folders with its host, docker containers can't see the contents of NFS shares.

**Comment by jarus** [comment]
In my eyes this is a important feature to share storage and data between containers. Is there a update?

**Comment by fardok** [comment]
Im have some problem with using  share folders from VirtualBox and NFS:

1. Start Centos7 in VirtualBox  - ok, its VirtualMachine
2. mount  vboxsf or nfs   from mainhost  192.168.0.1 to /SharedFolder  in VirtualMachine - ok, no problem, i see files inside
3.   docker run   somecontainer -v  /SharedFolder:/containerfolder   - ok, but THEY EMPTY!

So i dont understand how to use docker if impossible to  use nfs and vboxsf.  Any another protocols?

**Comment by fardok** [comment]
iscsi? ftp? v9fs?    i need to run  thousands containers but havent  disks 100Tb  on that machine

**Comment by fardok** [comment]
Solution:
chkconfig  docker off
service docker  stop
service  nfs----dont remember   stop then start
mount  nfs folder from somewhere
service  docker start

Understand?

**Comment by christianTragesser** [comment]
@fardok on CentOS 7 as the docker host platform, mounting nfs shares before starting the docker service successfully presents NFS shares to my volume containers. Thanks for the suggestion.

**Comment by guillaume-brossard-adsk** [comment]
+1 It would be really useful for us to be able to use the data container pattern with nfs.

**Comment by vladvetu** [comment]
@fardok Thanks. It's working on Fedora 21 Server.

**Comment by ajodock** [comment]
This is what I did to fix this issue on my RHEL/CentOS7 machines

    #Copy the default unit file to the override location
    cp /usr/lib/systemd/system/docker.service /etc/systemd/system/
    
    #Edit /etc/systemd/system/docker.service and replace
    #After=network.target docker.socket
    #with
    #After=network.target docker.socket remote-fs.target local-fs.target
    
    #Reload systemd settings
    systemctl daemon-reload
    
    #Restart docker
    systemctl restart docker

Technically the docker restart is all that is required to fix it, but the systemd unit will make sure that it works after a reboot as well, but beware that any mountpoints (local or remote) that you mount AFTER docker is already running will not be accessable to the docker container.  You have to restart the docker service every time you mount a new device/nfs share to be able to access it from a docker volume.

**Comment by tr3buchet** [comment]
@ajodock systemd service files should be extended using "drop-ins":

```
cat > /etc/systemd/system/docker.service.d/remote-fs.conf << EOF
[Unit]
# append remote-fs.target and local-fs.target to docker.service's After targets
After=remote-fs.target local-fs.target
EOF
systemctl daemon-reload
```
more info [here](https://fedoraproject.org/wiki/Systemd#How_do_I_customize_a_unit_file.2F_add_a_custom_unit_file.3F)
also check out `systemd-delta` and `systemd-unit` for more info

**Comment by thefotios** [comment]
Just to throw my hat in the ring here, removing "MountFlags=private" in my systemd service worked perfectly. This is on Fedora 21

**Comment by ajodock** [comment]
@tr3buchet Thanks!  Another reason to love systemd ;)

@thefotios Looks like it just got fixed in Fedora 21 on the 4th (https://bugzilla.redhat.com/show_bug.cgi?id=1185423)  Thanks for the info this is a much better fix.

**Comment by pikeas** [comment]
I'm experiencing a related issue, which I began describing at https://github.com/docker/docker/issues/10589.

In brief, I have an OS X machine, running a Vagrant CoreOS VM for docker, with my files mounted via NFS. When I `--volumes-from` mount a directory from OSX to Docker, the container is able to write to that folder for a period of time, but eventually locks up utterly. The container can't be `stop`ed, `kill`ed, or `rm -f`ed. The only solution is to `vagrant halt; vagrant up` and then `rm` it.

I don't have a one-liner for reproduction, but I can reproduce when working with moderate to large numbers of files:

    `docker run --rm -it -v `pwd`:/tmp/app my_image bash`
    # in container
    cp -R /tmp/app/* /app    # read succeeds
    cd /app
    # installs a bunch of packages from package.json in cwd
    npm install
    cp -R /app/node_modules /tmp/app
    # locks up partway through write

I'm running fully updated versions of OS X, Vagrant, CoreOS (alpha channel), and docker itself.

Please let me know if I should open a separate issue for this.

**Comment by axman** [comment]
I was having a similar problem with OS X and a CentOS 7 VM.  It turned out to be SE Linux related.  To tell SE Linux to get out of the way use.

```
setsebool -P virt_use_nfs on
```

This also fixed an issue where I could access the files by name but not view the directory listing.  You may have to restart docker for the change to take effect.

**Comment by jbemmel** [comment]
Had the same issue on an Ubuntu VM running Docker, due to a race condition at boot the NFS mount was not available if Docker started first. I tried mounting a volume /vm and then NFS mounting /vm/nfs, but that did not fix the issue. Have now added 'and mounted MOUNTPOINT=/vm/nfs' to /etc/init/docker.conf on start, which seems to work.

It would be nice if Docker volumes could accept changes to NFS mount points in their hierarchy though, a mount might become temporarily unavailable and I'm not sure if Docker can deal with that

**Comment by phani1kumar** [comment]
@jbemmel can you please explain the solution as to where to add the 'and mounted MOUNTPOINT=/gm/nfs'? Seems I am facing similar problem with Ubuntu. 
I've a ubuntu 14.04 guest inside a ubuntu 14.04 host machine running oracle virtualbox (4.3). 

I am mounting an nfs drive on the guest machine, which I am using as a docker volume. But the volume is not recognized by docker even after I restart the docker service. 

I am mounting nfs drive through:
```
mount -t nfs -o proto=tcp 192.168.1.101:/mnt/my_nas/nmedia/mysql /mnt/smedia_data
```
1) Stopped docker service before mounting and started it after mounting, --> volumes didn't work
2) restarted docker service after mounting `service docker restart` --> volumes didn't work

please help

**Comment by jbemmel** [comment]
Hi @phani1kumar,
It appears that NFS functionality depends on the underlying storage driver ( aufs, devicemapper, etc. ), as well as the sharing of file handles between processes ( see http://blog.hashbangbash.com/2014/11/docker-devicemapper-fix-for-device-or-resource-busy-ebusy/ ) i.e. 'unshare' may have an impact on NFS mounts.

I've moved away from using the 'MOUNTPOINT=/vm/nfs' as I am not sure if that event is even emitted. Instead I created an upstart file like this:
cat > /etc/init/ecdn.conf << EOF
description "eCDN container"
author "Jeroen van Bemmel"
# mounted MOUNTPOINT=/vm/nfs doesn't seem to work, at least not the first time
start on started docker and virtual-filesystems
stop on starting rc RUNLEVEL=[016]
respawn
script
  exec /usr/bin/docker start -a ecdn
end script

pre-stop script
        /usr/bin/docker stop ecdn
#  dont /usr/bin/docker rm ecdn
end script

EOF

and then create the container like this:
script -c "docker create -it --name='ecdn' --volume /vm:/usr/share/nginx/html/vm:ro image/name"

**Comment by amulyas** [comment]
+1 I am running into same issue .. this is open from 1.5 years is Docker planning to fix or add this functionality..  . or we need to use some work around ..

**Comment by cpuguy83** [comment]
Being that docker is not doing anything with the nfs mount, I do not see what would be done here.
Docker will bind-mount a dir from the host... if something breaks there it's likely a kernel level issue or some communication issue with the NFS server.

**Comment by LordFPL** [comment]
Hello, for information check my comment https://github.com/docker/docker/issues/4213#issuecomment-67670645 : it was linked to mount namespace.

In speed, i found this blog : https://huaminchen.wordpress.com/2015/05/19/how-docker-handles-mount-namespace/

Conclusion :  check your docker systemd properties... if you have a "MountFlags=private", delete it... if no, try to add "MountFlags=shared", and restart docker of course :)

**Comment by crosbymichael** [comment]
Don't add the volume ontop of the nfs mount point,  if you mounted nfs at `/mnt/nfs` don't add `-v /mnt/nfs:/data` as the volume.  Place the volume within the nfs mountpoint and it will work just fine.

`-v /mnt/nfs/somedata:/data`

**Comment by prologic** [comment]
I'm experiencing this problem too and I don't understand @crosbymichael 's suggestion.

See: https://gist.github.com/prologic/68077eedc1ecc3e0feda

Mounting the NFS share:
```#!bash
docker run -i -t --rm --cap-add SYS_ADMIN --name nfs -v /mnt/nfs -e SERVER=10.0.0.7 -e SHARE=/mnt/data/cache/packages prologic/nfs-client /bin/sh
# ls /mnt/nfs
... lots of files ...
```

Connecting another containers with ``--volumes-from``:
```#!bash
$ docker run -i -t --rm --volumes-from nfs alpine /bin/sh
/ # ls /mnt/nfs
/ #
```

The image: https://github.com/prologic/nfs-client

**Comment by ryansch** [comment]
I have this working.  See: https://github.com/outstand/docker-nfs-client/blob/master/rancheros-service.yml

**Comment by cpuguy83** [comment]
Closing since this is now possible by using `shared` or `rshared` when mounting a host dir to a container.
For example `docker run -v /foo:/bar:shared myNFSImage`, this would allow you to mount from inside the container to `/bar` and have that propagate to the host @ `/foo`

Thanks.

**Comment by flaccid** [comment]
Tested `:shared` with a `--volumes-from` container too it seems ok now on 1.10. Shameless plug but might be useful for others to use https://github.com/flaccid/docker-nfs-client.

**Comment by kiwenlau** [comment]
@cpuguy83

On **192.168.59.1**, I mount a nfs directory within **container A** using following command
```
sudo mount 192.168.59.1:/opt /opt
``` 

I want share /opt directory with other containers on **192.168.59.1**. Therefore, I start **container A** with "-v /opt:/opt:shared"

I get following errors...

```
docker: Error response from daemon: Cannot start container a978fe6271f3c26404a31fc2de871b6e2f25d99044f0905bfdb8165de0710f39: Path /opt is mounted on / but it is not a shared mount..```

Thank you!

**Comment by cpuguy83** [comment]
@kiwenlau `mount --bind /opt /opt && mount --make-shared /opt`
If that doesn't work I'm going to assume you are on systemd, in which case you need to change the MountFlags to shared.

**Comment by kiwenlau** [comment]
@cpuguy83 It works perfectly, thank you!

**Comment by prologic** [comment]
@cpuguy83 On RancherOS 0.4.5 here (Docker Engine 1.10.3): Couldn't find any other references to this but I'm pretty sure you'd run ``mount --bind /opt /opt && mount --make-shared /opt`` on the Docker host side right (not in the container)? Can you confirm? If so this makes sharing NFS mounts initiated by a container service responsible for mounting said NFS shares with other containers via ``--volumes-from`` somewhat inconvenient since you have to have some pre-host setup first?

**Comment by cpuguy83** [comment]
@prologic it depends on the host distro, but yes, it must be done on the host side.

**Comment by prologic** [comment]
@cpuguy83 Yeah that's what I thought (and tested and confirmed too!); also I did mention the host "RancherOS 0.4.5" :) -- Can you point me to docs regarding the ``-v /bar:/foo:shared``? (I haven't been able to find any :/)

**Comment by ryansch** [comment]
@prologic: The following should help!

RancherOS cloud config:
```
write_files:
  - path: /opt/rancher/bin/start.sh
    permissions: "0700"
    owner: root:root
    content: |
      #!/bin/bash
      mkdir /opt
      mount --bind /opt /opt
      mount --make-shared /opt
```

Volume docs: https://docs.docker.com/engine/reference/run/#volume-shared-filesystems

**Comment by prologic** [comment]
@ryansch Thanks Ryan! Unfortunately I have to modify the RancherOS config itself; ProxMox VE doesn't support cloud-config as such (yet?) :)

**Comment by ryansch** [comment]
@prologic As long as you create `start.sh` in that location through some means (ssh?) it will have the intended effect.  RancherOS runs that on each boot.
